### PR TITLE
Implementation of insert and eject media

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Reference:
https://github.com/terraform-providers/terraform-provider-vcd/issues/126

Add implementation for insert and eject media for VM. In others words this functionality allows to mount media(iso file) as cdrom in vm.
* works for system or org administrators
* eject possible only if vm powered off

Signed-off-by: Vaidotas Bauzys <vbauzys@vmware.com>

### CODE REVIEW

File: govcd/vm.go [hunk 1]
```diff
@@ -545,3 +557,134 @@ func (vm *VM) DetachDisk(diskParams *types.DiskAttachOrDetachParams) (Task, erro
...
+	}
+
+	if vmStatus != types.VAppStatuses[8] {
+		return Task{}, fmt.Errorf("to eject media, vm has to be power off state")
+	}
+
+	return vm.insertOrEjectMedia(mediaParams, types.RelMediaEjectMedia)
+}
+
+func validateMediaParams(mediaParams *types.MediaInsertOrEjectParams) error {
```
Comment (dataclouder) [inline govcd/vm.go]: Please add the function purpose

File: govcd/vm_test.go [hunk 2]
```diff
@@ -416,3 +416,146 @@ func (vcd *TestVCD) Test_VMDetachDisk(check *C) {
...
+	//verify
+	err = vm.Refresh()
+	check.Assert(err, IsNil)
+	check.Assert(isMediaInjected(vm.VM.VirtualHardwareSection.Item), Equals, false)
+}
+
+func isMediaInjected(items []*types.VirtualHardwareItem) bool {
+	isFound := false
+	for _, hardwareItem := range items {
+		if hardwareItem.ResourceSubType == "vmware.cdrom.iso" {
```
Comment (dataclouder) [inline govcd/vm_test.go]: Where does this literal value come from? Can use use a constant with a good name and documentation instead?

File: govcd/vm.go [hunk 3]
```diff
@@ -21,13 +21,25 @@ type VM struct {
 	client *Client
 }
 
+type VMRecord struct {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: Types is inside struct -> types.QueryResultVMRecordType and it has all details.

File: govcd/vm_test.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/vm_test.go]: No documentation found - I figure out from responses

File: govcd/vm.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/vm.go]: done

File: govcd/vm.go [hunk 4]
```diff
@@ -545,3 +557,134 @@ func (vm *VM) DetachDisk(diskParams *types.DiskAttachOrDetachParams) (Task, erro
...
+			Name: media.CatalogItem.Entity.Name,
+			ID:   media.CatalogItem.Entity.ID,
+			Type: media.CatalogItem.Entity.Type,
+		},
+	})
+
+	return task, err
+}
+
+// Helper function which finds media ant calls EjectMedia
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: done

File: govcd/vm.go [hunk 5]
```diff
@@ -545,3 +557,134 @@ func (vm *VM) DetachDisk(diskParams *types.DiskAttachOrDetachParams) (Task, erro
 
 	return vm.attachOrDetachDisk(diskParams, types.RelDiskDetach)
 }
+
+// Helper function which finds media ant calls InsertMedia
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: done

File: govcd/vdc_test.go [hunk 6]
```diff
@@ -196,3 +196,21 @@ func (vcd *TestVCD) Test_FindMediaImage(check *C) {
 	check.Assert(err, IsNil)
 	check.Assert(mediaImage, Equals, MediaItem{})
 }
+
+func (vcd *TestVCD) Test_QueryVM(check *C) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vdc_test.go]: done

File: govcd/media_test.go [hunk 7]
```diff
@@ -131,3 +131,33 @@ func (vcd *TestVCD) Test_DeleteMediaImage(check *C) {
 	check.Assert(mediaItem, Equals, MediaItem{})
 
 }
+
+// Tests System function FindMediaAsCatalogItem
+func (vcd *TestVCD) Test_FindMediaAsCatalogItem(check *C) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media_test.go]: done

File: govcd/vdc.go [hunk 8]
```diff
@@ -572,6 +572,43 @@ func (vdc *Vdc) FindVMByName(vapp VApp, vm string) (VM, error) {
 	return VM{}, fmt.Errorf("can't find vm: %s", vm)
 }
 
+func (vdc *Vdc) QueryVM(vappName, vmName string) (VMRecord, error) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vdc.go]: done

File: govcd/media_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media_test.go]: This just helper function for insert/eject media. Can be used later for other cases too.

File: govcd/media_test.go [hunk 9]
```diff
@@ -131,3 +131,34 @@ func (vcd *TestVCD) Test_DeleteMediaImage(check *C) {
 	check.Assert(mediaItem, Equals, MediaItem{})
 
 }
+
+// Tests System function FindMediaAsCatalogItem by creating media item and
+// and finding it as catalog item after.
+func (vcd *TestVCD) Test_FindMediaAsCatalogItem(check *C) {
+	skipWhenMediaPathMissing(vcd, check)
+	itemName := TestUploadMedia + "6"
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media_test.go]: Cause already 5 test created media with constant TestUploadMedia

File: govcd/vdc_test.go [hunk 10]
```diff
@@ -196,3 +196,23 @@ func (vcd *TestVCD) Test_FindMediaImage(check *C) {
 	check.Assert(err, IsNil)
 	check.Assert(mediaImage, Equals, MediaItem{})
 }
+
+// Tests function QueryVM by searching vm created
+// by test suite
+func (vcd *TestVCD) Test_QueryVM(check *C) {
+
+	if vcd.vapp.VApp == nil {
+		check.Skip("No Vapp provided")
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vdc_test.go]: fixed

File: govcd/vm_test.go [hunk 11]
```diff
@@ -416,3 +416,147 @@ func (vcd *TestVCD) Test_VMDetachDisk(check *C) {
...
+	check.Assert(err, IsNil)
+
+	//verify
+	err = vm.Refresh()
+	check.Assert(err, IsNil)
+	check.Assert(isMediaInjected(vm.VM.VirtualHardwareSection.Item), Equals, false)
+}
+
+// Test Insert or Eject Media for VM
+func (vcd *TestVCD) Test_InsertOrEjectMedia(check *C) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm_test.go]: Handle is wrapper with additional functionality which handles catalog and media queries and data type creation for calling further function.

File: govcd/media.go [hunk 12]
```diff
@@ -342,3 +342,23 @@ func (mediaItem *MediaItem) Delete() (Task, error) {
 	// The request was successful
 	return *task, nil
 }
+
+func FindMediaAsCatalogItem(org *Org, catalogName, mediaName string) (CatalogItem, error) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media.go]: done

File: govcd/vm.go [hunk 13]
```diff
@@ -21,13 +21,25 @@ type VM struct {
...
+}
+
 func NewVM(cli *Client) *VM {
 	return &VM{
 		VM:     new(types.VM),
 		client: cli,
 	}
 }
 
+func NewVMRecord(cli *Client) *VMRecord {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: done

File: govcd/vm.go [hunk 14]
```diff
@@ -545,3 +557,135 @@ func (vm *VM) DetachDisk(diskParams *types.DiskAttachOrDetachParams) (Task, erro
...
+	err := validateMediaParams(mediaParams)
+	if err != nil {
+		return Task{}, err
+	}
+
+	return vm.insertOrEjectMedia(mediaParams, types.RelMediaInsertMedia)
+}
+
+// Eject media from VM
+// Call insertOrEjectMedia with media and types.RelMediaEjectMedia to eject media from vm.
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: done

File: govcd/vm.go [hunk 15]
```diff
@@ -545,3 +557,135 @@ func (vm *VM) DetachDisk(diskParams *types.DiskAttachOrDetachParams) (Task, erro
...
+		return Task{}, err
+	}
+
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return Task{}, err
+	}
+
+	if vmStatus != types.VAppStatuses[8] {
+		return Task{}, fmt.Errorf("to eject media, vm has to be power off state")
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: done

**Pull request merged**